ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,435 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+                                    GuidString,
+                                    ReferenceString);
+    NT_ASSERT(NT_SUCCESS(Status));
 
-    if (DeviceKey != NULL)
-        DeviceKeyRealP = DeviceKey;
-    else
-        DeviceKeyRealP = &DeviceKeyReal;
+    DPRINT("New symbolic link is '%wZ'\n", &SymbolicLink);
+
+    *SymbolicLinkName = SymbolicLink;
+    return STATUS_SUCCESS;
+}
+
+static
+NTSTATUS
+IopSeparateSymbolicLink(
+    _In_ PCUNICODE_STRING SymbolicLinkName,
+    _Out_opt_ PUNICODE_STRING PrefixString,
+    _Out_opt_ PUNICODE_STRING MungedString,
+    _Out_opt_ PUNICODE_STRING GuidString,
+    _Out_opt_ PUNICODE_STRING ReferenceString,
+    _Out_opt_ PBOOLEAN ReferenceStringPresent,
+    _Out_opt_ LPGUID InterfaceClassGuid)
+{
+    UNICODE_STRING KernelModePrefix = RTL_CONSTANT_STRING(L"\\??\\");
+    UNICODE_STRING UserModePrefix = RTL_CONSTANT_STRING(L"\\\\?\\");
+    UNICODE_STRING MungedPrefix = RTL_CONSTANT_STRING(L"##?#");
+    UNICODE_STRING MungedStringReal, GuidStringReal, ReferenceStringReal;
+    UNICODE_STRING DeviceString;
+    UNICODE_STRING LinkNameNoPrefix;
+    USHORT i, ReferenceStringOffset;
+    NTSTATUS Status;
+
+    DPRINT("Symbolic link is '%wZ'\n", SymbolicLinkName);
+
+    /* Symbolic link name looks like \??\ACPI#PNP0501#1#{GUID}\ReferenceString */
+    /* Make sure it starts with the expected prefix */
+    if (!RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE) &&
+        !RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
+    {
+        DPRINT1("Invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Sanity checks */
+    ASSERT(KernelModePrefix.Length == UserModePrefix.Length);
+    ASSERT(SymbolicLinkName->Length >= KernelModePrefix.Length);
+
+    /* Make a version without the prefix for further processing */
+    LinkNameNoPrefix.Buffer = SymbolicLinkName->Buffer + KernelModePrefix.Length / sizeof(WCHAR);
+    LinkNameNoPrefix.Length = SymbolicLinkName->Length - KernelModePrefix.Length;
+    LinkNameNoPrefix.MaximumLength = LinkNameNoPrefix.Length;
+
+    DPRINT("Symbolic link without prefix is '%wZ'\n", &LinkNameNoPrefix);
 
-    if (InstanceKey != NULL)
-        InstanceKeyRealP = InstanceKey;
+    /* Find the reference string, if any */
+    for (i = 0; i < LinkNameNoPrefix.Length / sizeof(WCHAR); i++)
+    {
+        if (LinkNameNoPrefix.Buffer[i] == L'\\')
+        {
+            break;
+        }
+    }
+    ReferenceStringOffset = i * sizeof(WCHAR);
+
+    /* The GUID is before the reference string or at the end */
+    ASSERT(LinkNameNoPrefix.Length >= ReferenceStringOffset);
+    if (ReferenceStringOffset < GUID_STRING_BYTES + sizeof(WCHAR))
+    {
+        DPRINT1("Invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Get reference string (starts after {GUID} and \) from link without prefix */
+    // TODO: Add some more sanity checks
```

Comment:
Any ideas what some of these other sanity checks could be?

Reply:
From the start, I though to add the same checks as in the original code (as you can see, I've copied this part from `IoSetDeviceInterfaceState` at the bottom). But after that my code was modified many times. Looking at the current code now, I see that there is nothing to check here, because everything is already checked above (looks like I copied the checks from the original code to there, and in the original code, nothing is checked at this point). So I think this comment can be removed now. :slightly_smiling_face:

Reply:
Removed.